server/src/main/java/org/opensearch/cluster/metadata/IndexMetadata.java

```diff
@@ -634,6 +635,69 @@ public static APIBlock readFrom(StreamInput input) throws IOException {
         Property.IndexScope
     );
 
+    /**
+     * Used to specify the type for the ingestion index. If not specified, the ingestion source not enabled
+     */
+    public static final String SETTING_INGESTION_SOURCE_TYPE = "index.ingestion_source.type";
+    public static final String NONE_INGESTION_SOURCE_TYPE = "none";
+    public static final Setting<String> INGESTION_SOURCE_TYPE_SETTING = Setting.simpleString(
+        SETTING_INGESTION_SOURCE_TYPE,
+        NONE_INGESTION_SOURCE_TYPE,
+        new Setting.Validator<>() {
+
+            @Override
+            public void validate(final String value) {
+                // TODO: validate this with the registered types in the ingestion source plugin
+            }
+
+            @Override
+            public void validate(final String value, final Map<Setting<?>, Object> settings) {
+                // TODO: validate this with the ingestion source params
+            }
+        },
+        Property.IndexScope,
+        Property.Dynamic
+    );
+
+    /**
+     * Used to specify initial reset policy for the ingestion pointer. If not specified, default to the latest
+     */
+    public static final String SETTING_INGESTION_SOURCE_POINTER_INIT_RESET = "index.ingestion_source.pointer.init.reset";
+    public static final Setting<String> INGESTION_SOURCE_POINTER_INIT_RESET_SETTING = Setting.simpleString(
+        SETTING_INGESTION_SOURCE_POINTER_INIT_RESET,
+        StreamPoller.ResetState.LATEST.name(),
+        new Setting.Validator<>() {
+
+            @Override
+            public void validate(final String value) {
+                if (!(value.equalsIgnoreCase(StreamPoller.ResetState.LATEST.name())
+                    || value.equalsIgnoreCase(StreamPoller.ResetState.EARLIEST.name()))) {
+                    throw new IllegalArgumentException(
+                        "Invalid value for " + SETTING_INGESTION_SOURCE_POINTER_INIT_RESET + " [" + value + "]"
+                    );
+                }
+            }
+
+            @Override
+            public void validate(final String value, final Map<Setting<?>, Object> settings) {}
+        },
+        Property.IndexScope,
+        Property.Dynamic
+    );
+
+    public static final Setting.AffixSetting<Object> INGESTION_SOURCE_PARAMS_SETTING = Setting.prefixKeySetting(
+        "index.ingestion_source.param.",
+        key -> new Setting<>(key, "", (value) -> {
+            // TODO: add ingestion source params validation
+            return value;
+        }, Property.IndexScope)
+    );
```

Comment:
Using `prefixKeySetting(...)` for the ingest source-specific params means that we must be able to represent the settings as a flat set of key/value pairs (since, internally, all settings are converted to `key=value` representations, where `key` is a string and `value` is either a primitive type or a list of primitive types). 

This doesn't give us as much flexibility as JSON for these params. In particular, we don't have a good way of representing a list of objects via settings.

As long as we don't think an ingestion source will ever need to take a list (or array) of objects as part of its configuration, this should be fine. In practice, I think we *should* be okay.

Reply:
I believe it should be fine for Kinesis as well.

I recall that settings were a problem for search pipelines, but a search pipeline is fundamentally a list of processors, where each processor's config is an object.

I haven't been able to think of a problem that arises from using settings, even a day later. So, I think I'm comfortable saying that it's okay.